File: .tmp/demo_driver.py
"""Real-bank-grade demo letter generation — drive the full pipeline through API management.

Usage:
    python .tmp/demo_driver.py inspect                 # show current state
    python .tmp/demo_driver.py publish <externalId>    # advance template to PUBLISHED + policy
    python .tmp/demo_driver.py generate <externalId>   # create credential + runtime generate
"""
import json, os, sys, time, urllib.request, urllib.error

BASE = os.environ.get("BACKEND_URL", "http://localhost:8080")
MGMT = BASE + "/api/management/v1"
RUNTIME = BASE + "/api/dev/v1"

USERS = {
    "admin":      ("10000001", "ChangeMe123!"),
    "group_admin":("10000002", "ChangeMe123!"),
    "author":     ("10000003", "ChangeMe123!"),
}

def http(method, url, token=None, body=None, multipart=None, raw=False):
    headers = {}
    data = None
    if token: headers["Authorization"] = "Bearer " + token
    if multipart:
        # multipart via http.client is fiddly; shell out to curl
        return curl_multipart(method, url, headers, multipart)
    if body is not None:
        data = json.dumps(body).encode()
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=data, method=method, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            raw_bytes = r.read()
            if raw: return r.status, dict(r.headers), raw_bytes
            if not raw_bytes: return r.status, None, None
            try: return r.status, dict(r.headers), json.loads(raw_bytes)
            except: return r.status, dict(r.headers), raw_bytes.decode("utf-8","replace")
    except urllib.error.HTTPError as e:
        raw_bytes = e.read()
        try: err = json.loads(raw_bytes)
        except: err = raw_bytes.decode("utf-8","replace")
        return e.code, dict(e.headers), err

def curl_multipart(method, url, headers, fields):
    import subprocess
    args = ["curl","-sS","-X",method,url]
    for k,v in headers.items():
        args += ["-H", f"{k}: {v}"]
    for k,v in fields.items():
        if isinstance(v, tuple):  # (filepath)
            args += ["-F", f"{k}=@{v[0]}"]
        else:
            args += ["-F", f"{k}={v}"]
    r = subprocess.run(args, capture_output=True, text=True)
    try: return 200, {}, json.loads(r.stdout)
    except: return 0, {}, r.stdout

def login(role="group_admin"):
    u,p = USERS[role]
    s,_,b = http("POST", MGMT+"/auth/login", body={"username":u,"password":p})
    assert s==200 and b, f"login {role} failed: {s} {b}"
    return b["result"]["accessToken"]

def get_json(path, token):
    s,h,b = http("GET", MGMT+path, token=token)
    return s, b

def main():
    cmd = sys.argv[1] if len(sys.argv)>1 else "inspect"
    if cmd=="inspect":
        ga = login("group_admin")
        s, b = get_json("/templates?size=200", ga)
        print(f"templates (status {s}):")
        if s==200:
            for t in b["result"]:
                print(f"  {t.get('externalId','?'):32} | {t.get('lifecycleStatus','?'):14} | {t.get('groupCode','?'):8} | {t.get('name','')[:50]}")
        else:
            print(b)
        s, b = get_json("/masters?size=200", ga)
        print(f"masters (status {s}):")
        if s==200:
            for m in b["result"]:
                print(f"  {m.get('name','?'):48} | {m.get('reviewState','?'):10} | {m.get('groupCode','?'):8}")
        else:
            print(b)
    elif cmd=="generate":
        ext = sys.argv[2]
        ga = login("group_admin")
        # find template
        s,b = get_json("/templates?size=200", ga)
        tpl = next((t for t in b["result"] if t["externalId"]==ext), None)
        assert tpl, f"template {ext} not found"
        tid = tpl["id"]
        print(f"template: {ext} id={tid} status={tpl.get('lifecycleStatus')}")
        # create credential
        s,_,b = http("POST", MGMT+f"/templates/{tid}/api/credentials", token=ga)
        print(f"credential create: {s} {b}")
        cred = b["result"]
        # generate via runtime
        headers = {
            "X-Api-Credential-Id": cred["externalId"],
            "X-Api-Credential-Secret": cred["secret"],
            "X-Access-Account": "demo-real-letter-caller",
            "Content-Type": "application/json",
        }
        body = {
            "output": {"format":"DOCX","mode":"SYNC_STREAM"},
            "variables": {"customerName":"Northgate Manufacturing Ltd."} if "CREDIT" in ext else {"borrowerName":"Mr Oliver Hartley"},
            "requestId": f"req-demo-{int(time.time())}",
            "idempotencyKey": f"idem-demo-{int(time.time())}",
        }
        req = urllib.request.Request(f"{RUNTIME}/templates/{ext}/default/generate",
            data=json.dumps(body).encode(), method="POST", headers=headers)
        try:
            with urllib.request.urlopen(req, timeout=120) as r:
                doc = r.headers.get("documentId")
                print(f"GENERATE status={r.status} documentId={doc}")
                print("headers:", {k:v for k,v in r.headers.items()})
                if doc:
                    with open(f".tmp/generated_{ext}.docx","wb") as f: f.write(r.read())
                    print("saved .tmp/generated_"+ext+".docx")
        except urllib.error.HTTPError as e:
            print("GENERATE FAILED", e.code, e.read()[:500])

if __name__=="__main__":
    main()

File: .tmp/demo_driver.js
// Real-bank-grade demo driver — inspect / generate via the full pipeline.
// Usage: node .tmp/demo_driver.js inspect
//        node .tmp/demo_driver.js generate <externalId> <outFile>
const http = require('http');
const fs = require('fs');
const path = require('path');

const BASE = process.env.BACKEND_URL || 'http://127.0.0.1:8080';
const MGMT = BASE + '/api/management/v1';
const RUNTIME = BASE + '/api/dev/v1';

const USERS = {
  admin: ['10000001', 'ChangeMe123!'],
  group_admin: ['10000002', 'ChangeMe123!'],
  author: ['10000003', 'ChangeMe123!'],
};

function resultList(body) {
  if (!body) return [];
  if (Array.isArray(body.result)) return body.result;
  if (body.result && Array.isArray(body.result.content)) return body.result.content;
  if (Array.isArray(body.content)) return body.content;
  return [];
}

function request(method, urlStr, { token, body, headers: extraHeaders = {}, raw = false } = {}) {
  return new Promise((resolve, reject) => {
    const u = new URL(urlStr);
    const bodyData = body != null ? JSON.stringify(body) : null;
    const headers = { ...(extraHeaders || {}) };
    if (token) headers['Authorization'] = 'Bearer ' + token;
    if (bodyData) { headers['Content-Type'] = 'application/json'; headers['Content-Length'] = Buffer.byteLength(bodyData); }
    const req = http.request({
      method, protocol: u.protocol, hostname: u.hostname, port: u.port, path: u.pathname + u.search,
      headers, timeout: 120000,
    }, (res) => {
      const chunks = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => {
        const buf = Buffer.concat(chunks);
        if (raw) return resolve({ status: res.statusCode, headers: res.headers, body: buf });
        const text = buf.toString('utf8');
        let json = null;
        try { json = text ? JSON.parse(text) : null; } catch (e) { json = text; }
        resolve({ status: res.statusCode, headers: res.headers, body: json });
      });
    });
    req.on('error', reject);
    req.on('timeout', () => { req.destroy(new Error('timeout')); });
    if (bodyData) req.write(bodyData);
    req.end();
  });
}

async function login(role) {
  const [u, p] = USERS[role];
  const r = await request('POST', MGMT + '/auth/login', { body: { username: u, password: p } });
  if (r.status !== 200) throw new Error(`login ${role} failed: ${r.status} ${JSON.stringify(r.body)}`);
  return r.body.result.accessToken;
}

async function main() {
  const [, , cmd, ...rest] = process.argv;
  const action = cmd || 'inspect';
  if (action === 'inspect') {
    const tok = await login('group_admin');
    const tpl = await request('GET', MGMT + '/templates?size=200', { token: tok });
    console.log('=== TEMPLATES ===');
    const tpls = resultList(tpl.body);
    if (tpl.status === 200) {
      for (const t of tpls) {
        console.log(`${(t.externalId||'?').padEnd(32)} | ${(t.lifecycleStatus||'?').padEnd(14)} | ${(t.groupCode||'?').padEnd(8)} | ${(t.name||'').slice(0,50)}`);
      }
    } else { console.log(tpl.body); }
    const m = await request('GET', MGMT + '/masters?size=200', { token: tok });
    console.log('=== MASTERS ===');
    const masters = resultList(m.body);
    if (m.status === 200) {
      for (const x of masters) {
        console.log(`${(x.name||'').padEnd(50)} | ${(x.reviewState||'NONE').padEnd(10)} | ${x.groupCode||'?'}`);
      }
    } else { console.log(m.body); }
  } else if (action === 'generate') {
    const ext = rest[0];
    const outFile = rest[1] || `.tmp/generated_${ext}.docx`;
    const ga = await login('group_admin');
    const tpl = await request('GET', MGMT + '/templates?size=200', { token: ga });
    const target = resultList(tpl.body).find((t) => t.externalId === ext);
    if (!target) throw new Error('template ' + ext + ' not found');
    const tid = target.id;
    console.log(`template: ${ext} id=${tid} status=${target.lifecycleStatus}`);
    const cred = await request('POST', MGMT + `/templates/${tid}/api/credentials`, { token: ga });
    console.log('credential externalId=' + cred.body.result.externalId);
    const headers = {
      'X-Api-Credential-Id': cred.body.result.externalId,
      'X-Api-Credential-Secret': cred.body.result.secret,
      'X-Access-Account': 'svc-caller',
    };
    let vars = { customerName: 'Executive Sample Customer' };
    if (ext === 'DEMO-CREDIT-LIMIT-CONFIRM') {
      vars = { customerName:'Northgate Manufacturing Ltd.', facilityReference:'CORP-CL-2026-88421', creditLimit:'GBP 12,500,000.00', reviewDate:'2026-06-30', includeOverdraft:true, overdraftLimit:'GBP 1,000,000.00' };
    } else if (ext === 'DEMO-MORTGAGE-APPROVAL') {
      vars = { borrowerName:'Mr Oliver Hartley & Mrs Sarah Hartley', propertyAddress:'14 Willow Close, Bristol BS8 4QT', loanAmount:'GBP 385,000.00', interestRate:'4.29% fixed for 5 years', termYears:25, approvalDate:'2026-06-20' };
    }
    const body = {
      output: { format: 'DOCX', mode: 'SYNC_STREAM' },
      variables: vars,
      requestId: 'req-demo-' + Date.now(),
      idempotencyKey: 'idem-demo-' + Date.now(),
    };
    const r = await request('POST', RUNTIME + `/templates/${ext}/default/generate`, { headers, body, raw: true });
    console.log('GENERATE status=' + r.status);
    console.log('documentId=' + (r.headers['documentid'] || r.headers['documentId']));
    if (r.status === 200 && r.body && r.body.length) {
      fs.writeFileSync(outFile, r.body);
      console.log(`saved ${outFile} (${r.body.length} bytes)`);
    } else {
      console.log('body:', r.body.toString('utf8').slice(0, 800));
    }
  } else {
    console.log('Unknown action:', action);
  }
}

main().catch((e) => { console.error('ERROR:', e.message); process.exit(1); });
